File: C++/design-a-stack-with-increment-operation.cpp
class CustomStack {
public:
    vector<int> stack;
    int maxSz;
    
    CustomStack(int maxSize) {
        maxSz = maxSize;
    }
    
    void push(int x) {
        if (stack.size() < maxSz) {
            stack.push_back(x);
        }
    }
    
    int pop() {
        int x = -1;
        if (stack.size() > 0) {
            x = stack.back();
            stack.pop_back();
        }
        return x;
    }
    
    void increment(int k, int val) {
        int sz = min((int)stack.size(), k);
        for (int i = 0; i < sz; i++) {
            stack[i] += val;
        }
    }
};

/**
 * Your CustomStack object will be instantiated and called as such:
 * CustomStack* obj = new CustomStack(maxSize);
 * obj->push(x);
 * int param_2 = obj->pop();
 * obj->increment(k,val);
 */

File: Go/design-a-stack-with-increment-operation.go
type CustomStack struct {
	stack   []int
	size    int
}

func Constructor(maxSize int) CustomStack {
	return CustomStack{
		make([]int, maxSize),
		0,
	}
}

func (this *CustomStack) Push(x int) {
	if this.size >= len(this.stack) {
		return
	}
	this.stack[this.size] = x
	this.size++
}

func (this *CustomStack) Pop() int {
	if this.size <= 0 {
		return -1
	}
	this.size--
	top := this.stack[this.size]
	return top
}

func (this *CustomStack) Increment(k int, val int) {
	for i := 0; i < k && i < len(this.stack); i++ {
		this.stack[i] += val
	}
}

/**
 * Your CustomStack object will be instantiated and called as such:
 * obj := Constructor(maxSize);
 * obj.Push(x);
 * param_2 := obj.Pop();
 * obj.Increment(k,val);
 */